test/powershell/Language/Scripting/PipelineBehaviour.Tests.ps1.bak

```diff
@@ -0,0 +1,122 @@
+
+        BeforeAll {
+            $powershell = $null
+        }
+
+        AfterEach {
+            if ($powershell) {
+                $powershell.Dispose()
+                $powershell = $null
+            }
+        }
+
+        It 'does not execute End {} if the pipeline is halted during Process {}' {
+            # We don't need Should -Not -Throw as if this reaches end{} and throws the test will fail anyway.
+            1..10 |
+                & {
+                    begin { "BEGIN" }
+                    process { "PROCESS $_" }
+                    end { "END"; throw "This should not be reached." }
+                } |
+                Select-Object -First 3 |
+                Should -Be @( "BEGIN", "PROCESS 1", "PROCESS 2" )
+        }
+
+        It 'still executes Cleanup {} if the pipeline is halted' {
+            $Script = {
+                1..10 |
+                    & {
+                        process { $_ }
+                        cleanup { throw "Cleanup block hit." }
+                } |
+                    Select-Object -First 1
+            }
+            $Script | Should -Throw -ExpectedMessage "Cleanup block hit."
+        }
+
+        It 'still executes Cleanup {} when StopProcessing() is triggered mid-pipeline' {
+            $script = @"
+                function test {
+                    begin {}
+                    process {
+                        Start-Sleep -Seconds 10
+                    }
+                    end {}
+                    cleanup {
+                        Write-Information "CLEANUP"
+                    }
+
+                }
+"@
+            $powershell = [powershell]::Create()
+            $powershell.AddScript($script).AddStatement().AddCommand('test') > $null
+
+            $asyncResult = $powershell.BeginInvoke()
+            Start-Sleep -Seconds 2
+            $powershell.Stop()
+
+            $powershell.EndInvoke($asyncResult) > $null
+            $powershell.Streams.Information[0].MessageData | Should -BeExactly "CLEANUP"
+        }
+
+        It 'still completes Cleanup {} execution when StopProcessing() is triggered mid-Cleanup {}' {
+            $script = @"
+                function test {
+                    begin {}
+                    process {
+                        "PROCESS"
+                    }
+                    end {}
+                    cleanup {
+                        Start-Sleep -Seconds 10
+                        Write-Information "CLEANUP"
+                    }
+
+                }
+"@
+            $powershell = [powershell]::Create()
+            $powershell.AddScript($script).AddStatement().AddCommand('test') > $null
+
+            $asyncResult = $powershell.BeginInvoke()
+            Start-Sleep -Seconds 2
+            $powershell.Stop()
+
+            $output = $powershell.EndInvoke($asyncResult)
+
+            $output | Should -Be "PROCESS"
+            $powershell.Streams.Information[0].MessageData | Should -BeExactly "CLEANUP"
```

Comment:
should probably add some checks of the other streams here and a check of the `$powershell.HadErrors` property (and perhaps some explicit tests for errors, etc)

Reply:
Absolutely. The tests are from the original PR #9900, and I disabled it for now because there is behavior difference in my alternative design. I will make sure add adequate tests for the error handling once the proposal is accepted by the committee.

Reply:
Lots of tests were added, especially error handling tests.